### ISSUE TITLE
FDDividendAmericanEngine - incorrect results

### ISSUE THREAD

**Issue by pstrus** [amaggiulli/QLNet#211]
Valuation of a American option with a discrete dividend using FDDividendAmericanEngine yields incorrect results (and different to Quantlib).

I was trying to value a following option:
Valuation date: 20 July 2018
Maturity date: 17 Aug 2018
Type: Call
Spot: 2900
Strike: 2800
Volatility : 20%
Interest rate : 0%

Dividend (paid one day before expiry)
Date : 16 Aug 2018
Value: 40

The option value produced by QLNet is
**97.057**
Similar valuation using Quantlib returns
**124.308**

Now when I change the dividend to 0 both QLNet and Quantlib return same number:
**125.20**

Since the dividend is paid just one day before expiry and it's a american option I would expect the option price to be close to 124.27 which is a price of an european option expiring one day earlier (16 Aug)

I tried changing dividend date (for example to 22 July 2018) and the price returned is always same:
97.057, which suggests that the model is actually treating the option as a european one (instead of american)

## Your Environment
* Version used: QLNet 1.11.1 (I tried on develop branch as well)

**Comment by amaggiulli** [comment]
Hi Jakub  , 
thank you for spotting this issue , I will create a feature now with a test case with your data .
Just one question :  the Quantlib value ( 124.308 ) is incorrect too ?

**Comment by amaggiulli** [comment]
Also , can you share the code that produce 97.057 using QLNet  ?

**Comment by pstrus** [comment]
Hi Andrea,

I think the value produced by Quantlib is correct.

The code to produce the **95.057** is almost exactly as in the test you added. The only difference is that I used different number of timeSteps and gridPoints when initializing FDDividendAmericanEngine. I used 1000, instead of default, which is 100.
`var pricingEngine = new FDDividendAmericanEngine(process, 1000, 1000);`

When I change back to 100, both QLNet and Quantlib return the numbers you get.

I looked at the code and found where the problem is.
FDMultiPeriodEngine class overrides initializeStepCondition() method:
```
      protected override void initializeStepCondition()
      {
         stepCondition_ = new NullCondition<Vector>();
      }
```

The problem is that the "AmericanCondition" is set on a base class FDConditionEngineTemplate, and since it's overridden by the derived class it's actually never accessed (and valuation behaves like for European option).

One way to solve is to remove the method override from FDMultiPeriodEngine and change initializeStepCondition() method in FDConditionEngineTemplate 
from:
```
      protected virtual void initializeStepCondition()
      {
         if (stepConditionImpl_ == null)
            throw new NotSupportedException();
         else
            stepCondition_ = stepConditionImpl_();
      }
```
To:
```
      protected virtual void initializeStepCondition()
      {
         if (stepConditionImpl_ == null)
            stepCondition_ = new NullCondition<Vector>();
         else
            stepCondition_ = stepConditionImpl_();
      }

```
That way the base class will default to NullCondition in case there is no stepConditionImpl_ set.

I did that locally and now FDAmericanDividendEngine results do match Quantlib.
Also all unit tests do pass, which is promising, however I have not explicitly checked any other models and I'm not sure if this solution is generic enough and if it follows what has been done in Quantlib, so I'll leave the rest to you :).

Hope this helps.
Jakub

**Comment by amaggiulli** [comment]
Thank you Jakub ,

for now this is a good solution , at first look seems to me that the c# mimic the c++ code structure but not the logic. This can happen when porting  C++ templates in CSharp. Numbers are correct now , maybe a refactoring will follow. Thank you for finding the bug and proposing the solution :)

~Andrea

**Comment by amaggiulli** [comment]
Fixed in release 1.11.2
